### PR TITLE
Support force tokens to % of total experts during calibration

### PR DESCRIPTION
**Pull request by cjluo-nv**
## What does this PR do?

**Type of change:** New feature

**Overview:** Adds a configurable `moe_calib_experts_ratio` parameter that controls the percentage of experts to calibrate during the forward pass in MoE (Mixture of Experts) models. Previously, the calibration forward always routed tokens to **all** experts, which is expensive. This PR allows the user to specify a ratio (default: still all experts so no behavior change) to improve expert calibration coverage without the cost of a full-expert forward. The token counting for the expert coverage table now tracks the calibration routing and runs on CUDA for efficiency.

**Changes include:**
- New `moe_calib_experts_ratio` field in `QuantizeAlgorithmConfig` (`config.py`)
- Propagation of the ratio from the algorithm config to MoE modules during calibration (`mode.py`)
- Updated `_QuantSparseMoe.forward` to use the configurable ratio instead of hard-coding all experts (`huggingface.py`)
- New `--moe_calib_experts_ratio` CLI flag in `hf_ptq.py` (default `0.25`)
- Moved `expert_token_count` tensor to CUDA and updated the HTML table title in `moe_utils.py`

## Usage

Via hf_ptq.py CLI — calibrate 50% of experts during MoE calibration
python hf_ptq.py --model <model> --qformat int4_awq --moe_calib_experts_ratio 0.5

Via Python API — pass the ratio through the algorithm config
import modelopt.torch.quantization as mtq

quant_cfg = {
    "quant_cfg": { ... },
    "algorithm": {
        "method": "awq_lite",
        "moe_calib_experts_ratio": 0.25,  # calibrate 1/4 of experts
    },
}
mtq.quantize(model, quant_cfg, forward_loop=calib_loop)

## Testing
Test with Qwen3 30B A3B calibration and check the tokens per expert.

## Release Notes

* **New Features**
  * Added support for configurable expert calibration during Mixture of Experts (MOE) model quantization. Users can now specify the percentage of experts to include during calibration, enabling better expert coverage and improved quantization accuracy for MOE models. Default: 25% of all experts.

### CODE REVIEW

**Comment by cjluo-nv** [comment]
Data for Qwen3 30B:

[moe_before.html](https://github.com/user-attachments/files/25432206/moe_before.html) -- original HF forward, no force routing
[moe_after_2.html](https://github.com/user-attachments/files/25432209/moe_after_2.html) -- 1/8 routing ratio
[moe_after_3.html](https://github.com/user-attachments/files/25432210/moe_after_3.html) -- 1/4 routing ratio

File: modelopt/torch/quantization/plugins/huggingface.py [hunk 1]
```diff
@@ -483,36 +486,50 @@ def _gate_forward_hook(self, module, input, output):
                 logits = output if not isinstance(output, tuple) else output[0]
                 top_k = self.gate.top_k if hasattr(self.gate, "top_k") else self.top_k
                 _, indices = torch.topk(logits.float(), top_k, dim=-1)
-            counts = torch.bincount(
-                indices.reshape(-1).cpu(), minlength=len(self.expert_token_count)
-            )
-            self.expert_token_count += counts
+            counts = torch.bincount(indices.reshape(-1), minlength=len(self.expert_token_count))
```
Comment (realAsma) [inline modelopt/torch/quantization/plugins/huggingface.py]: len call also triggers CPU-GPU sync  in my understanding
```suggestion
            counts = torch.bincount(indices.reshape(-1), minlength=self.expert_token_count.shape[0])
```

File: modelopt/torch/quantization/plugins/huggingface.py [hunk 2]
```diff
@@ -483,36 +484,50 @@ def _gate_forward_hook(self, module, input, output):
...
             if TRANSFORMERS_VERSION_GE_5_0:
                 assert hasattr(self, "gate") and hasattr(self.gate, "top_k")
                 original_top_k = self.gate.top_k
-                self.gate.top_k = self.gate.num_experts
+                self.gate.top_k = round(self.gate.num_experts * self._moe_calib_experts_ratio)
+                assert self.gate.top_k >= original_top_k, (
+                    f"moe_calib_experts_ratio {self._moe_calib_experts_ratio},"
+                    f" calib top_k {self.gate.top_k} smaller than original"
+                    f" top_k {original_top_k}"
+                )
```
[the remark this answers is not in the document]
Comment (realAsma) [reply modelopt/torch/quantization/plugins/huggingface.py]: +1

File: modelopt/torch/quantization/config.py [hunk 3]
```diff
@@ -1091,6 +1070,16 @@ class QuantizeAlgorithmConfig(ModeloptBaseConfig):
         title="This field specifies the name of the calibration algorithm. If None, no calibration is performed.",
     )
 
+    moe_calib_experts_ratio: float | None = ModeloptField(
```
Comment (realAsma) [inline modelopt/torch/quantization/config.py:1094]: Yea this is good idea to put this here.

File: examples/llm_ptq/example_utils.py [hunk 4]
```diff
@@ -232,6 +233,15 @@ def build_quant_cfg(
...
         )
 
+    if moe_calib_experts_ratio:
+        if isinstance(quant_cfg["algorithm"], str):
+            quant_cfg["algorithm"] = {
+                "method": quant_cfg["algorithm"],
+                "moe_calib_experts_ratio": moe_calib_experts_ratio,
+            }
+        else:
+            quant_cfg["algorithm"]["moe_calib_experts_ratio"] = moe_calib_experts_ratio
```
[the remark this answers is not in the document]
Comment (realAsma) [reply examples/llm_ptq/example_utils.py:244]: +1

**Review by realAsma — APPROVED** [review]
Overall looks great! I left my minor comments.

File: examples/llm_ptq/hf_ptq.py [hunk 5]
```diff
@@ -1126,6 +1127,15 @@ def parse_args() -> argparse.Namespace:
             "(sensitivity scores, costs, etc.). Only used when auto_quantize_bits is specified."
         ),
     )
+    parser.add_argument(
```
Comment (meenchen) [inline examples/llm_ptq/hf_ptq.py:1130]: Should we add something similar to mcore PTQ as well? cc @realAsma @jenchen13 @ChenhanYu

**Comment by Edwardf0t1** [comment]
@cjluo-nv could we try a larger MoE model, like Kimi-k2.5 or Qwen3.5-MoE with this change? I feel the accuracy impact  might be bigger for large MoE models with less MoE calibrated. 

Also, if 25% experts is calibrated, can we ensure all experts have a valid weight_scale and input_scale in the exported checkpoint?

File: examples/llm_ptq/hf_ptq.py [hunk 6]
```diff
@@ -1126,6 +1127,15 @@ def parse_args() -> argparse.Namespace:
...
+        "--moe_calib_experts_ratio",
+        type=float,
+        default=1.0 / 4,
```
Comment (Edwardf0t1) [inline examples/llm_ptq/hf_ptq.py]: How the 25% is selected?

File: examples/llm_ptq/example_utils.py [hunk 7]
```diff
@@ -201,6 +201,7 @@ def build_quant_cfg(
     model_type,
     quant_cfg_choices,
     kv_quant_cfg_choices,
+    moe_calib_experts_ratio,
```
[the remark this answers is not in the document]
Comment (jenchen13) [reply examples/llm_ptq/example_utils.py]: +1 can this argument be optional? 

btw have you run any experiments to see if there is accuracy degradation with forcing a % of experts during calibration? I recall previously some of us were concerned there would be accuracy degradation if we force non-routed experts to process tokens

File: examples/llm_ptq/hf_ptq.py (hunk 6 above)
Comment (jenchen13) [inline examples/llm_ptq/hf_ptq.py]: how was the default value chosen?

File: modelopt/torch/quantization/config.py (hunk 3 above)
Comment (jenchen13) [reply modelopt/torch/quantization/config.py:1094]: Is it necessary to put `moe_calib_experts_ratio` in the quant config? adding this to the quant config makes it seem like this field is very essential to PTQ but in reality it should be optional. Especially given the fact that there doesn't seem to be many experiments on the effect of this parameter on PTQ overall and whether it is an essential argument

Comment (jenchen13) [reply modelopt/torch/quantization/config.py:1094]: for example if we think `"moe_calib_experts_ratio"` is necessary to change the Quant config API, then why would we not include KV cache clamping as part of the quant config too or other non-essential PTQ features?

File: modelopt/torch/quantization/plugins/huggingface.py [hunk 8]
```diff
@@ -483,36 +486,47 @@ def _gate_forward_hook(self, module, input, output):
...
                 super().forward(hidden_states)
                 self.gate.top_k = original_top_k
             else:
                 # Path for transformers < 5.0
                 original_top_k = self.top_k
                 if hasattr(self, "num_experts"):
-                    self.top_k = self.num_experts
+                    self.top_k = max(
+                        original_top_k, round(self.num_experts * self._moe_calib_experts_ratio)
+                    )
```
[the remark this answers is not in the document]
Comment (jenchen13) [reply modelopt/torch/quantization/plugins/huggingface.py:520]: +1 ^ seems like a risky area for silent failure if rounding goes wrong

File: modelopt/torch/quantization/config.py (hunk 3 above)
Comment (cjluo-nv) [reply modelopt/torch/quantization/config.py:1094]: @jenchen13 this is a good question: why would we not include KV cache clamping as part of the quant config too or other non-essential PTQ features?

If we've decided to make this tunable, we probably also want to add it here.

File: examples/llm_ptq/hf_ptq.py (hunk 6 above)
Comment (cjluo-nv) [reply examples/llm_ptq/hf_ptq.py]: This is based on the qwen3 30B result and also balancing the PTQ efficiency. Baiscally:

Higher ratio: potentially more experts visited, longer calibration time.
Lower ration: potentially less experts visited, less calibration time.

Regards to accuracy:
In DS3.2, we've seen forwarding tokens to all experts has better accuracy. In some other MOEs @meenchen mentioned original forward has better accuracy.

This number has to be tuned.

If this is a concern, I will set the default to 1.0 so no behavior change and make an separate PR if we want to set this to a different number.

File: examples/llm_ptq/example_utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (cjluo-nv) [reply examples/llm_ptq/example_utils.py]: it can be model specific.

File: modelopt/torch/quantization/config.py (hunk 3 above)
Comment (realAsma) [reply modelopt/torch/quantization/config.py:1094]: +1 on making the KVCache clamping here - this will help the QAD to be aligned with export path

Comment (realAsma) [reply modelopt/torch/quantization/config.py:1094]: @jenchen13  this is a great idea
